WordPress/src/main/java/org/wordpress/android/ui/prefs/SettingsFragment.java

```diff
@@ -365,15 +376,64 @@ public boolean onPreferenceClick(Preference preference) {
         }
     };
 
+    private final OnPreferenceClickListener languagePreferenceClickListener = new OnPreferenceClickListener() {
+        @Override
+        public boolean onPreferenceClick(Preference preference) {
+            AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(getActivity());
+            dialogBuilder.setTitle(getString(R.string.language));
+
+            final String[] availableLocales = getResources().getStringArray(R.array.available_languages);
+            final String[] values = new String[availableLocales.length + 1];
+
+            values[0] = "Device" + " (" + Locale.getDefault().getDisplayLanguage() + ")";
+            for (int i = 0; i < availableLocales.length; ++i) {
+                Locale locale = new Locale(availableLocales[i]);
+                values[i + 1] = locale.getDisplayLanguage() + " (" + availableLocales[i] + ")";
+            }
+
+            ArrayAdapter<String> adapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_list_item_1, values);
+            ListView listView = new ListView(getActivity());
+            listView.setAdapter(adapter);
+            listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
+                @Override
+                public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+                    Resources res = getResources();
+                    DisplayMetrics dm = res.getDisplayMetrics();
+                    Configuration conf = res.getConfiguration();
+                    conf.locale = new Locale(values[position]);
+                    res.updateConfiguration(conf, dm);
+
+                    mSettings.edit().putString(SETTINGS_PREFERENCES, availableLocales[position]).apply();
```

Comment:
Off by one error here: `position` can be `availableLocales.length + 1` (from line 391) `availableLocales[availableLocales.length + 1]  will crash`